core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusOAuth2Client.java

```diff
@@ -0,0 +1,351 @@
+import com.nimbusds.openid.connect.sdk.validators.InvalidHashException;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.hash.Hashing.sha256;
+import static com.nimbusds.oauth2.sdk.ResponseType.CODE;
+import static com.nimbusds.openid.connect.sdk.OIDCScopeValue.OPENID;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class NimbusOAuth2Client
+        implements OAuth2Client
+{
+    private static final Logger LOG = Logger.get(NimbusAirliftHttpClient.class);
+
+    private final Issuer issuer;
+    private final ClientID clientId;
+    private final ClientSecretBasic clientAuth;
+    private final Scope scope;
+    private final String principalField;
+    private final Set<String> accessTokenAudiences;
+    private final Duration maxClockSkew;
+    private final NimbusHttpClient httpClient;
+    private final Provider<OAuth2ServerConfiguration> serverConfigurationProvider;
+    private volatile boolean loaded;
+    private URI authUrl;
+    private URI tokenUrl;
+    private Optional<URI> userinfoUrl;
+    private JWSKeySelector<SecurityContext> jwsKeySelector;
+    private JWTProcessor<SecurityContext> accessTokenProcessor;
+    private AuthorizationCodeFlow flow;
+
+    @Inject
+    public NimbusOAuth2Client(OAuth2Config oauthConfig, Provider<OAuth2ServerConfiguration> serverConfigurationProvider, NimbusHttpClient httpClient)
+    {
+        requireNonNull(oauthConfig, "oauthConfig is null");
+        issuer = new Issuer(oauthConfig.getIssuer());
+        clientId = new ClientID(oauthConfig.getClientId());
+        clientAuth = new ClientSecretBasic(clientId, new Secret(oauthConfig.getClientSecret()));
+        scope = Scope.parse(oauthConfig.getScopes());
+        principalField = oauthConfig.getPrincipalField();
+        maxClockSkew = oauthConfig.getMaxClockSkew();
+
+        accessTokenAudiences = new HashSet<>(oauthConfig.getAdditionalAudiences());
+        accessTokenAudiences.add(clientId.getValue());
+        accessTokenAudiences.add(null); // A null value in the set allows JWTs with no audience
+
+        this.serverConfigurationProvider = requireNonNull(serverConfigurationProvider, "serverConfigurationProvider is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+    }
+
+    @Override
+    public void load()
```

Comment:
this method is only used in Server context. Why is it not a matter of an implementation of the interface? I would suggest adding `@AfterConstruct` annotation from javax on this method and removing it `load` from the interface.

Reply:
The idea here is to load the class outside the guice context similarly to many other loadable components.